keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+    // to the ResourceClaim after it has been created.
+   Template *ResourceClaimTemplate
+}
+
+type ResourceClaimTemplate struct {
+    // May contain labels and annotations that will be copied into the ResourceClaim
+    // when creating it. No other fields are allowed and will be rejected during
+    // validation.
+    metav1.ObjectMeta
+
+    // The specification for the ResourceClaim. The entire content is
+    // copied unchanged into the ResourceClaim that gets created from this
+    // template. The same fields as in a ResourceClaim
+    // are also valid here.
+    Spec ResourceClaimSpec
+}
+```
+
+### kube-controller-manager
+
+The code that creates a ResourceClaim from an inline ResourceClaimTemplate will
+be an almost verbatim copy of the [generic ephemeral volume
+code](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral),
+just with different types.
+
+kube-controller-manager will need new [RBAC
+permissions](https://github.com/kubernetes/kubernetes/commit/ff3e5e06a79bc69ad3d7ccedd277542b6712514b#diff-2ad93af2302076e0bdb5c7a4ebe68dd3188eee8959c72832181a7597417cd196) that allow creating ResourceClaims.
+
+As a future extension, kube-controller-manager could also remove ReservedBy
+entries that reference deleted objects. This could serve as fallback for cases
+where the entity that normal removes those entries was unable to do so, for
+example kubelet on a node that shut down unexpectedly.
+
+### kube-scheduler
+
+The scheduler plugin needs to implement several extension points. The [volume
+binder
+plugin](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/framework/plugins/volumebinding)
+can serve as a reference.
+
+#### Pre-filter
+
+This checks whether a Pod uses any ResourceClaims. If there are ResourceClaims
+with immediate binding that are not allocated yet, then the Pod will be marked
+as unschedulable at the moment.
+
+#### Filter
+
+This checks whether the given node has access to those ResourceClaims which
+were already allocated.
+
+For unallocated ResourceClaims with delayed allocation, only those nodes are
+filtered out that are explicitly listed in UnsuitableNodes. There are several
+reasons why such a deny list is more suitable than an allow list:
+- Nodes for which no information is available must pass the filter phase to be
+  included in the list that will be passed to the post-filter and to get copied
+  into the PotentialNodes field there.
+- A node can already be chosen while there is no information yet and, if
+  allocation for that node actually works, the Pod can get scheduled sooner.
+- Some resource drivers might not have any unsuitable nodes, for example
+  because they modify containers and that works on all nodes at all
+  times. Forcing such drivers to set an allow list would cause unnecessary
+  work.
+
+In its state for the Pod the scheduler plugin must remember when it rejected a
```

Comment:
will the Filter always return Unschedulable in a first pass?

Reply:
No, the opposite: if all other resources (RAM, CPU) are sufficient and there's no explicit deny entry for node in the ResourceClaim, Filter will proceed with the node.

In other words, an allocation attempt and a request to provide information about suitable nodes get kicked off in parallel. If the allocation succeeds, great! If not, then the next pass will have more information and hopefully make a better choice.

Reply:
During the initial pass, Filter will let a node through although the allocation is not complete yet. The scheduling pass then stops in the Reserve operation because an unallocated ResourceClaim cannot be reserved.